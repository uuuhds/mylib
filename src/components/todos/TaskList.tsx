import React from 'react';
import { TaskEvents, TaskProps, Task } from './';
import PropTypes from 'prop-types';
import { achiveTask, pinTask, selectTasks } from '../../lib/reducers/todoSlice';
import { useSelector, useDispatch } from 'react-redux';

export type TaskListProps = {
  loading: boolean;
  tasks: TaskProps[];
} & TaskEvents;

export function TaskList() {
  const tasks = useSelector(selectTasks);
  const dispatch = useDispatch();
  return (
    <PureTaskList
      tasks={tasks}
      onArchiveTask={(id: string) => dispatch(achiveTask(id))}
      onPinTask={(id: string) => dispatch(pinTask(id))}
    />
  );
}

export function PureTaskList({
  loading,
  tasks,
  onPinTask,
  onArchiveTask,
}: TaskListProps) {
  /* 以前的 TaskList 实现 */
  const events = {
    onPinTask,
    onArchiveTask,
  };

  const LoadingRow = (
    <div className="loading-item">
      <span className="glow-checkbox" />
      <span className="glow-text">
        <span>Loading</span> <span>cool</span> <span>state</span>
      </span>
    </div>
  );
  if (loading) {
    return (
      <div className="list-items">
        {LoadingRow}
        {LoadingRow}
        {LoadingRow}
        {LoadingRow}
        {LoadingRow}
        {LoadingRow}
      </div>
    );
  }
  if (tasks.length === 0) {
    return (
      <div className="list-items">
        <div className="wrapper-message">
          <span className="icon-check" />
          <div className="title-message">You have no tasks</div>
          <div className="subtitle-message">Sit back and relax</div>
        </div>
      </div>
    );
  }
  const tasksInOrder = [
    ...tasks.filter(t => t.state === 'TASK_PINNED'),
    ...tasks.filter(t => t.state !== 'TASK_PINNED'),
  ];
  return (
    <div className="list-items">
      {tasksInOrder.map(task => (
        <Task key={task.id} task={task} {...events} />
      ))}
    </div>
  );
}

PureTaskList.propTypes = {
  /** Checks if it's in loading state */
  loading: PropTypes.bool,
  /** The list of tasks */
  tasks: PropTypes.arrayOf(Task.propTypes.task).isRequired,
  /** Event to change the task to pinned */
  onPinTask: PropTypes.func.isRequired,
  /** Event to change the task to archived */
  onArchiveTask: PropTypes.func.isRequired,
};

PureTaskList.defaultProps = {
  loading: false,
};
