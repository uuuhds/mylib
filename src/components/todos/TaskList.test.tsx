import React from 'react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { WithPinnedTasks } from './TaskList.stories';
import { TaskListProps } from './';

it('renders pinned tasks at the start of the list', () => {
  const onArchiveTask = jest.fn();
  const onPinTask = jest.fn();
  const props = WithPinnedTasks.args as TaskListProps;
  const component = renderer.create(
    <WithPinnedTasks
      {...props}
      onPinTask={onPinTask}
      onArchiveTask={onArchiveTask}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
