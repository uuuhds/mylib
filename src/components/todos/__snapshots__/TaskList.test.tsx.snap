// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders pinned tasks at the start of the list 1`] = `
<div
  className="list-items"
>
  <div
    className="list-item TASK_PINNED"
  >
    <label
      className="checkbox"
    >
      <input
        defaultChecked={false}
        disabled={true}
        name="checked"
        type="checkbox"
      />
      <span
        className="checkbox-custom"
        onClick={[Function]}
      />
    </label>
    <div
      className="title"
    >
      <input
        placeholder="Input title"
        readOnly={true}
        type="text"
        value="Task 6 (pinned)"
      />
    </div>
    <div
      className="actions"
      onClick={[Function]}
    >
      <a
        onClick={[Function]}
      >
        <span
          className="icon-star"
        />
      </a>
    </div>
  </div>
  <div
    className="list-item TASK_INBOX"
  >
    <label
      className="checkbox"
    >
      <input
        defaultChecked={false}
        disabled={true}
        name="checked"
        type="checkbox"
      />
      <span
        className="checkbox-custom"
        onClick={[Function]}
      />
    </label>
    <div
      className="title"
    >
      <input
        placeholder="Input title"
        readOnly={true}
        type="text"
        value="Task 1"
      />
    </div>
    <div
      className="actions"
      onClick={[Function]}
    >
      <a
        onClick={[Function]}
      >
        <span
          className="icon-star"
        />
      </a>
    </div>
  </div>
  <div
    className="list-item TASK_INBOX"
  >
    <label
      className="checkbox"
    >
      <input
        defaultChecked={false}
        disabled={true}
        name="checked"
        type="checkbox"
      />
      <span
        className="checkbox-custom"
        onClick={[Function]}
      />
    </label>
    <div
      className="title"
    >
      <input
        placeholder="Input title"
        readOnly={true}
        type="text"
        value="Task 2"
      />
    </div>
    <div
      className="actions"
      onClick={[Function]}
    >
      <a
        onClick={[Function]}
      >
        <span
          className="icon-star"
        />
      </a>
    </div>
  </div>
  <div
    className="list-item TASK_INBOX"
  >
    <label
      className="checkbox"
    >
      <input
        defaultChecked={false}
        disabled={true}
        name="checked"
        type="checkbox"
      />
      <span
        className="checkbox-custom"
        onClick={[Function]}
      />
    </label>
    <div
      className="title"
    >
      <input
        placeholder="Input title"
        readOnly={true}
        type="text"
        value="Task 3"
      />
    </div>
    <div
      className="actions"
      onClick={[Function]}
    >
      <a
        onClick={[Function]}
      >
        <span
          className="icon-star"
        />
      </a>
    </div>
  </div>
  <div
    className="list-item TASK_INBOX"
  >
    <label
      className="checkbox"
    >
      <input
        defaultChecked={false}
        disabled={true}
        name="checked"
        type="checkbox"
      />
      <span
        className="checkbox-custom"
        onClick={[Function]}
      />
    </label>
    <div
      className="title"
    >
      <input
        placeholder="Input title"
        readOnly={true}
        type="text"
        value="Task 4"
      />
    </div>
    <div
      className="actions"
      onClick={[Function]}
    >
      <a
        onClick={[Function]}
      >
        <span
          className="icon-star"
        />
      </a>
    </div>
  </div>
  <div
    className="list-item TASK_INBOX"
  >
    <label
      className="checkbox"
    >
      <input
        defaultChecked={false}
        disabled={true}
        name="checked"
        type="checkbox"
      />
      <span
        className="checkbox-custom"
        onClick={[Function]}
      />
    </label>
    <div
      className="title"
    >
      <input
        placeholder="Input title"
        readOnly={true}
        type="text"
        value="Task 5"
      />
    </div>
    <div
      className="actions"
      onClick={[Function]}
    >
      <a
        onClick={[Function]}
      >
        <span
          className="icon-star"
        />
      </a>
    </div>
  </div>
</div>
`;
